crates/typst/src/layout/grid/layout.rs

```diff
@@ -677,39 +954,155 @@ impl<'a> GridLayouter<'a> {
+                    hline_stroke_at_column,
+                )
+                .into_iter()
+                .map(|(stroke, dx, length)| {
+                    let stroke = (*stroke).clone().unwrap_or_default();
+                    let thickness = stroke.thickness;
+                    let half = thickness / 2.0;
+                    let dx = if self.is_rtl { self.width - dx - length } else { dx };
+                    let target = Point::with_x(length + thickness);
+                    let hline = Geometry::Line(target).stroked(stroke);
+                    (
+                        thickness,
+                        Point::new(dx - half, dy),
                         FrameItem::Shape(hline, self.span),
-                    );
-                }
+                    )
+                });
 
-                // Render vertical lines.
-                for (x, dx) in points(self.rcols.iter().copied()).enumerate() {
-                    let dx = if self.is_rtl { self.width - dx } else { dx };
-                    // We want each vline to span the entire table (start
-                    // at y = 0, end after all rows).
-                    // We use 'split_vline' to split the vline such that it
-                    // is not drawn above colspans.
-                    for (dy, length) in
-                        split_vline(self.grid, rows, x, 0, self.grid.rows.len())
-                    {
-                        let target = Point::with_y(length + thickness);
-                        let vline = Geometry::Line(target).stroked(stroke.clone());
-                        frame.prepend(
-                            Point::new(dx, dy - half),
-                            FrameItem::Shape(vline, self.span),
-                        );
-                    }
-                }
+                // Draw later (after we sort all lines below.)
+                lines.extend(segments);
+            }
+
+            // Sort by increasing thickness, so that we draw larger strokes
+            // on top.
+            // This avoids layering problems where a smaller hline appears
+            // "inside" a larger vline. When both have the same size, hlines
+            // are drawn on top (since the sort is stable, and they are
+            // pushed first).
+            lines.sort_by_key(|(thickness, ..)| *thickness);
+
+            // We render lines in the reverse order because we are prepending
+            // them, meaning the first prepended line will be on the top of the
+            // others, the second will be below the first one, and so on; and,
+            // yet, the lines vector is sorted from bottom layer to top layer,
+            // so we have to prepend from its end to its start.
+            for (_, point, shape) in lines.into_iter().rev() {
+                // Prepend lines to the frame so they always appear below
+                // cells' content.
+                frame.prepend(point, shape);
             }
```

Comment:
> Preprending in a loop will be inefficient because prepend just does insert(0, ..) into a vector, so we'll get O(n^2) runtime. Better to use prepend_multiple here (takes an iterator).

Thanks for the heads-up!

> Let's maybe also move this down and push the cell backgrounds into the vector, too, because they also currently call prepend in a loop.

This would require calling `reverse()` on the stroke vector before pushing the fills (otherwise the fills would go above strokes). I'm not sure how much cheaper that is compared to calling `prepend_multiple` twice... probably at least a bit though since it won't affect anything else in the frame.

Or we could just sort the vector with both fills and strokes and have fills always go first.

I'm going with the `reverse()` approach for now, as it sounds more straightforward, but tell me what you think.